atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/AbortingCommitTsLoader.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2018 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.sweep;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.cache.CacheLoader;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Streams;
+import com.palantir.atlasdb.AtlasDbConstants;
+import com.palantir.atlasdb.keyvalue.api.KeyAlreadyExistsException;
+import com.palantir.atlasdb.transaction.impl.TransactionConstants;
+import com.palantir.atlasdb.transaction.service.TransactionService;
+import com.palantir.logsafe.SafeArg;
+
+public class AbortingCommitTsLoader extends CacheLoader<Long, Long> {
+    private static final Logger log = LoggerFactory.getLogger(AbortingCommitTsLoader.class);
+    private final TransactionService transactionService;
+
+    public AbortingCommitTsLoader(TransactionService transactionService) {
+        this.transactionService = transactionService;
+    }
+
+    @Override
+    public Long load(Long startTs) {
+        Optional<Long> maybeCommitTs = tryGetFromTransactionService(startTs);
+
+        if (!maybeCommitTs.isPresent()) {
+            maybeCommitTs = tryToAbort(startTs);
+        }
+
+        if (!maybeCommitTs.isPresent()) {
+            maybeCommitTs = tryGetFromTransactionService(startTs);
+        }
+
+        return maybeCommitTs.orElse(TransactionConstants.FAILED_COMMIT_TS);
+    }
+
+    @Override
+    public Map<Long, Long> loadAll(Iterable<? extends Long> nonCachedKeys) {
```

Comment:
This is weird - Long is final

Reply:
Looks like this is the `CacheLoader` implementation forcing this

Reply:
Yep, kind of awkward...